daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.speedy
+
+import java.util
+
+import com.daml.lf.PureCompiledPackages
+import com.daml.lf.data
+import com.daml.lf.language.Ast
+import com.daml.lf.language.Ast.{Expr, Package}
+import com.daml.lf.speedy.Compiler.FullStackTrace
+import com.daml.lf.speedy.SResult.{SResultFinalValue}
+import com.daml.lf.testing.parser.Implicits._
+import com.daml.lf.validation.Validation
+import org.scalatest.prop.TableDrivenPropertyChecks
+import org.scalatest.{Assertion, WordSpec, Matchers}
+
+class TailCallTest extends WordSpec with Matchers with TableDrivenPropertyChecks {
+
+  val pkg: Package =
+    p"""
+       module F {
+
+         // *Non* tail-recursive definition
+         val triangle : (Int64 -> Int64) = \ (x: Int64) ->
+           case (EQUAL @Int64 x 0) of
+               True -> 0
+             | _    -> ADD_INT64 x (F:triangle (SUB_INT64 x 1));
+
+         // Tail-recursive definition, via helper function
+         val triangleTR : (Int64 -> Int64) = F:triangleTR_acc 0;
+
+         // Tail-recursive definition using accumulator parameter
+         val triangleTR_acc : (Int64 -> Int64 -> Int64) = \ (acc: Int64) (x: Int64) ->
+           case (EQUAL @Int64 x 0) of
+               True -> acc
+             | _    -> F:triangleTR_acc (ADD_INT64 acc x) (SUB_INT64 x 1);
+
+
+         val triangle_viaFoldLeft : (Int64 -> Int64) = \ (x: Int64) ->
+            FOLDL @Int64 @Int64 ADD_INT64 0 (F:generate Nil@Int64 x);
+
+         val triangle_viaFoldRight : (Int64 -> Int64) = \ (x: Int64) ->
+            FOLDR @Int64 @Int64 ADD_INT64 0 (F:generate Nil@Int64 x);
+
+         // tail-recursive generator
+         val generate : (List Int64 -> Int64 -> List Int64) = \ (acc: List Int64) (x: Int64) ->
+           case (EQUAL @Int64 x 0) of
+               True -> acc
+             | _    -> F:generate (Cons @Int64 [x] acc) (SUB_INT64 x 1);
+
+       }
+      """
+
+  val small: Option[Int] = Some(5)
+  val unbounded: Option[Int] = None
+
+  "A *non* tail-recursive definition requires an unbounded env-stack, and an unbounded kont-stack" in {
+    val exp = e"F:triangle 100"
+    val expected = SValue.SInt64(5050)
+    // The point of this test is to prove that the bounded-evaluation checking really works.
+    runExpr(exp, envBound = unbounded, kontBound = unbounded) shouldBe expected
+    expectThrows {
+      runExpr(exp, envBound = small, kontBound = unbounded)
+    }
```

Comment:
Let's be a bit more precise on what gets thrown:
```suggestion
val error = the[RuntimeException] thrownBy runExpr(exp, envBound = small, kontBound = unbounded)
error.toString should include("BoundedArrayList, exceeded bound of 5")
```